saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationProvider.java

```diff
@@ -332,6 +336,48 @@ public void setResponseTimeValidationSkew(Duration responseTimeValidationSkew) {
 		this.responseTimeValidationSkew = responseTimeValidationSkew;
 	}
 
+	/**
+	 * Sets the assertion response custom decrypter.
+	 *
+	 * You can use this method like so:
+	 *
+	 * <pre>
+	 *	YourDecrypter decrypter = // ... your custom decrypter
+	 *
+	 *	OpenSamlAuthenticationProvider provider = new OpenSamlAuthenticationProvider();
+	 *	provider.setAssertionDecrypter((response) -> {
+	 *  	EncryptedAssertion encrypted = response.getEncryptedAssertions().get(0);
+	 *  	Assertion assertion = decrypter.decrypt(encrypted);
+	 *  	response.getAssertions().add(assertion);
+	 *	});
+	 * </pre>
+	 * @param assertionDecrypter response consumer
+	 */
+	public void setAssertionDecrypter(Consumer<Response> assertionDecrypter) {
```

Comment:
In the class, there is a tuple class called `ResponseToken` which has the deserialized SAML Response as well as the `Saml2AuthenticationToken`. It would be nice to have both of these when customizing decryption. The reason for that is if you are doing local decryption, you'll need to look up the keys from the `Saml2AuthenticationToken`.

Could we change this contract to `setAssertionDecrypter(Consumer<ResponseToken> assertionDecrypter)`?

Reply:
Agreed and done in the latest commit.